File: javascript/27.remove-element.js
/*
 * @lc app=leetcode id=27 lang=javascript
 *
 * [27] Remove Element
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} val
 * @return {number}
 */
var removeElement = function (nums, val) {
  // Solution 1:
  // let i = 0;
  // while (i < nums.length) {
  //   if (nums[i] === val) nums.splice(i, 1);
  //   else i++;
  // }
  // return nums.length;

  // Solution 2: slow and fast pointers
  let slow = 0,
    fast = 0;
  while (fast < nums.length) {
    if (nums[fast] !== val) {
      let temp = nums[slow];
      nums[slow++] = nums[fast];
      nums[fast] = temp;
    }
    fast++;
  }
  return slow;
};
// @lc code=end

File: rust/27.remove-element.rs
/*
 * @lc app=leetcode id=27 lang=rust
 *
 * [27] Remove Element
 */

// @lc code=start
impl Solution {
    pub fn remove_element(nums: &mut Vec<i32>, val: i32) -> i32 {
        // Solution 1:
        // nums.retain(|&num| num != val);
        // return nums.len() as i32;

        // Solution 2: two pointers
        let (mut s, mut f) = (0, 0);
        while f < nums.len() {
            if nums[f] != val {
                nums.swap(s, f);
                s += 1;
            }
            f += 1;
        }
        return s as i32;
    }
}
// @lc code=end
